File: FuncionesRPG.java
public class FuncionesRPG {

    // Retorna el daño real (mínimo 1)
    public static int calcularDano(int ataque, int defensa) {
        int dano = ataque - defensa;
        return dano > 0 ? dano : 1;
    }

    // Cura sin pasar del máximo
    public static double aplicarCuracion(double vida, double curacion, double maxVida) {
        double nuevaVida = vida + curacion;
        return nuevaVida > maxVida ? maxVida : nuevaVida;
    }

    // Muestra el estado del personaje
    public static void mostrarEstado(String nombre, double vida, int nivel) {
        System.out.printf("%s [Nv%d] HP: %.0f%n", nombre, nivel, vida);
    }

    // ===== EJERCICIO 4A =====
    // Sube de nivel si tiene suficiente XP
    public static int subirNivel(int xpActual, int xpNecesario, int nivelActual) {

        if (xpActual >= xpNecesario) {
            nivelActual++;
            xpActual = 0;

            System.out.println("¡Nivel alcanzado!");
            System.out.println("Nuevo nivel: " + nivelActual);
            System.out.println("XP reiniciado a: " + xpActual);
        } else {
            System.out.println("No hay suficiente XP para subir nivel.");
        }

        return nivelActual;
    }

    public static void main(String[] args) {

        System.out.println("===== PRUEBA calcularDano =====");
        int dano = calcularDano(20, 8);
        System.out.println("Daño realizado: " + dano);

        System.out.println("\n===== PRUEBA aplicarCuracion =====");
        double vidaCurada = aplicarCuracion(40, 80, 100);
        System.out.println("Vida después de curación: " + vidaCurada);

        System.out.println("\n===== PRUEBA mostrarEstado =====");
        mostrarEstado("Frodo", vidaCurada, 3);

        System.out.println("\n===== PRUEBA subirNivel (110 XP) =====");
        int nuevoNivel1 = subirNivel(110, 100, 3);
        System.out.println("Nivel retornado: " + nuevoNivel1);

        System.out.println("\n===== PRUEBA subirNivel (80 XP) =====");
        int nuevoNivel2 = subirNivel(80, 100, 3);
        System.out.println("Nivel retornado: " + nuevoNivel2);
    }
}

File: FuncionesRPG.py
# BLOQUE 4 - Funciones

def calcular_dano(ataque: int, defensa: int) -> int:

    dano = ataque - defensa

    return dano if dano > 0 else 1


def aplicar_curacion(
    vida: float,
    cur: float,
    max_vida: float
) -> float:

    nueva = vida + cur

    return min(nueva, max_vida)


def mostrar_estado(
    nombre: str,
    vida: float,
    nivel: int
):

    print(f'{nombre} [Nv{nivel}] HP: {vida:.0f}')


# Pruebas

d = calcular_dano(20, 8)
print(f'Dano: {d}')

v = aplicar_curacion(40, 80, 100)

mostrar_estado('Frodo', v, 3)


# Ejercicio 4A

def subir_nivel(
    xp_actual: int,
    xp_necesario: int,
    nivel_actual: int
) -> int:

    if xp_actual >= xp_necesario:

        nivel_actual += 1
        xp_actual = 0

        print(f'Nivel alcanzado: {nivel_actual}')

    return nivel_actual


# Pruebas

nivel1 = subir_nivel(110, 100, 3)
print(f'Nuevo nivel: {nivel1}')

nivel2 = subir_nivel(80, 100, 3)
print(f'Nuevo nivel: {nivel2}')
